packages/oauth/src/install-provider.spec.js

```diff
@@ -612,22 +670,118 @@ describe('InstallProvider', async () => {
+        },
+        url: `http://example.com?state=${fakeState}&code=${fakeCode}`,
+      };
+      await installer.handleCallback(req, res, successExpectedCallbackOptions);
       assert.isTrue(sent);
       assert.equal(fakeStateStore.verifyStateParam.callCount, 0);
     });
   });
+
+  it('should fail if the state value is not in cookies', async () => {
+    const fakeStateStore = {
+      generateStateParam: sinon.fake.resolves('fakeState'),
+      verifyStateParam: sinon.fake.resolves({})
+    };
+    let sent = false;
+    const res = {
+      send: () => { sent = true; },
+      setHeader: () => {},
+    };
+    const installer = new InstallProvider({ clientId, clientSecret, stateSecret, installationStore, stateStore: fakeStateStore, authVersion: 'v2', logger: noopLogger, clientOptions: fooClientOptions });
+    const fakeState = 'fakeState';
+    const fakeCode = 'fakeCode';
+    const req = {
+      headers: {
+        host: 'example.com',
+        // cookie: `slack-app-oauth-state=${fakeState}`,
+      },
+      url: `http://example.com?state=${fakeState}&code=${fakeCode}`,
+    };
+    await installer.handleCallback(req, res, failureExpectedCallbackOptions);
+    assert.isTrue(sent);
+  });
+  it('should fail if there is a different state value in cookies', async () => {
+    const fakeStateStore = {
+      generateStateParam: sinon.fake.resolves('fakeState'),
+      verifyStateParam: sinon.fake.resolves({})
+    };
+    let sent = false;
+    const res = {
+      send: () => { sent = true; },
+      setHeader: () => {},
+    };
+    const installer = new InstallProvider({ clientId, clientSecret, stateSecret, installationStore, stateStore: fakeStateStore, authVersion: 'v2', logger: noopLogger, clientOptions: fooClientOptions });
+    const fakeState = 'fakeState';
+    const fakeCode = 'fakeCode';
+    const req = {
+      headers: {
+        host: 'example.com',
+        cookie: `slack-app-oauth-state=something-different`,
+      },
+      url: `http://example.com?state=${fakeState}&code=${fakeCode}`,
+    };
+    await installer.handleCallback(req, res, failureExpectedCallbackOptions);
+    assert.isTrue(sent);
+  });
+  it('should not fail if legacyStateVerification is enabled', async () => {
+    const fakeStateStore = {
+      generateStateParam: sinon.fake.resolves('fakeState'),
+      verifyStateParam: sinon.fake.resolves({})
+    };
+    let sent = false;
+    const res = {
+      send: () => { sent = true; },
+      setHeader: () => {},
+    };
+    const installer = new InstallProvider({
+      clientId,
+      clientSecret,
+      stateSecret,
+      installationStore,
+      logger: noopLogger,
+      stateStore: fakeStateStore,
+      clientOptions: fooClientOptions,
+      legacyStateVerification: true,
+    });
+    const fakeState = 'fakeState';
+    const fakeCode = 'fakeCode';
+    const req = {
+      headers: {
+        host: 'example.com',
+        // cookie: `slack-app-oauth-state=${fakeState}`,
```

Comment:
Do we also want to check the case where legacyStateVerification=true and state cookie != state stored in state store?